feature/portlist/portlist.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package portlist contains code to poll the local system for open ports
+// and report them to the control plane, if enabled on the tailnet.
+package portlist
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"tailscale.com/envknob"
+	"tailscale.com/ipn"
+	"tailscale.com/ipn/ipnext"
+	"tailscale.com/ipn/ipnlocal"
+	"tailscale.com/ipn/policy"
+	"tailscale.com/portlist"
+	"tailscale.com/tailcfg"
+	"tailscale.com/types/logger"
+	"tailscale.com/util/eventbus"
+	"tailscale.com/version"
+)
+
+func init() {
+	ipnext.RegisterExtension("portlist", newExtension)
+}
+
+func newExtension(logf logger.Logf, sb ipnext.SafeBackend) (ipnext.Extension, error) {
+	busClient := sb.Sys().Bus.Get().Client("portlist")
+	e := &Extension{
+		sb:        sb,
+		busClient: busClient,
+		logf:      logger.WithPrefix(logf, "portlist: "),
+		pub:       eventbus.Publish[ipnlocal.PortlistServices](busClient),
+	}
+	e.ctx, e.ctxCancel = context.WithCancel(context.Background())
+	return e, nil
+}
+
+// Extension implements the portlist extension.
+type Extension struct {
+	ctx        context.Context
+	ctxCancel  context.CancelFunc
+	busClient  *eventbus.Client
+	pub        *eventbus.Publisher[ipnlocal.PortlistServices]
+	logf       logger.Logf
+	sb         ipnext.SafeBackend
+	stateStore ipn.StateStore
+	host       ipnext.Host // from Init
+
+	shouldUploadServicesAtomic atomic.Bool
+	startOnce                  sync.Once
+}
+
+func (e *Extension) Name() string { return "portlist" }
+func (e *Extension) Shutdown() error {
+	e.ctxCancel()
+	e.pub.Close()
+	e.busClient.Close()
+	return nil
+}
+
+func (e *Extension) Init(h ipnext.Host) error {
+	e.host = h
+	h.Hooks().ShouldUploadServices.Set(e.shouldUploadServicesAtomic.Load)
+	h.Hooks().BackendStateChange.Add(e.backendStateChange)
+	h.Hooks().ProfileStateChange.Add(e.profileStateChange)
+	return nil
+}
+
+func (e *Extension) backendStateChange(newState ipn.State) {
+	if newState != ipn.NoState {
+		return
+	}
+	e.startOnce.Do(e.readPoller)
+}
+
+func (e *Extension) profileStateChange(_ ipn.LoginProfileView, prefs ipn.PrefsView, sameNode bool) {
+	e.shouldUploadServicesAtomic.Store(e.shouldUploadServices(prefs))
+}
+
+func (e *Extension) shouldUploadServices(prefs ipn.PrefsView) bool {
+	if !prefs.Valid() || prefs.ShieldsUp() {
+		return false
+	}
+	nb := e.host.NodeBackend()
+	if nb == nil {
+		return false
+	}
+	return nb.CollectServices()
```

Comment:
The state change callback always receives valid prefs
https://github.com/tailscale/tailscale/blob/a40f23ad4a851d20abb6d339db3b82b8c6567a26/ipn/ipnext/ipnext.go#L313-L315
and `NodeBackend` is never `nil`.

So we can simplify this if we want:
```suggestion
	return !prefs.ShieldsUp() && e.host.NodeBackend().CollectServices()
```